src/iterators/mod.rs

```diff
@@ -1207,6 +1207,26 @@ fn chunk_iter_parts<A, D: Dimension>(
     (iter, n_whole_chunks, last_dim)
 }
 
+fn split_chunk_iter<A, D: Dimension>(
+    iter: AxisIterCore<A, D>,
+    n_whole_chunks: usize,
+    last_dim: D,
+    index: usize,
+) -> (
+    (AxisIterCore<A, D>, usize, D),
+    (AxisIterCore<A, D>, usize, D),
+) {
+    let left_n_whole_chunks = index;
```

Comment:
This implementation isn't quite right in the case where the axis is not evenly divisible by the chunk size and the specified `index` is equal to the number of chunks. Here's a test that fails with this implementation:

```rust
#[test]
fn axis_chunks_split_at() {
    let mut a = Array2::<usize>::zeros((11, 3));
    a.iter_mut().enumerate().for_each(|(i, elt)| *elt = i);
    for source in &[
        a.slice(s![..1, ..]),
        a.slice(s![..5, ..]),
        a.slice(s![..10, ..]),
        a.slice(s![..11, ..]),
    ] {
        let chunks_iter = source.axis_chunks_iter(Axis(0), 5);
        let all_chunks: Vec<_> = chunks_iter.clone().collect();
        let n_chunks = chunks_iter.len();
        assert_eq!(n_chunks, all_chunks.len());
        for index in 0..=n_chunks {
            let (left, right) = chunks_iter.clone().split_at(index);
            assert_eq!(&all_chunks[..index], &left.collect::<Vec<_>>()[..]);
            assert_eq!(&all_chunks[index..], &right.collect::<Vec<_>>()[..]);
        }
        assert_panics!({
            chunks_iter.split_at(n_chunks + 1);
        });
    }
}
```

One way to fix the implementation is this:

```rust
fn split_chunk_iter<A, D: Dimension>(
    iter: AxisIterCore<A, D>,
    n_whole_chunks: usize,
    last_dim: D,
    index: usize,
) -> (
    (AxisIterCore<A, D>, usize, D),
    (AxisIterCore<A, D>, usize, D),
) {
    // Note: `index` is checked to be `<= iter.len` in `iter.split_at(index)`.
    if index > n_whole_chunks {
        // In this case, the entire iterator stays in the left piece; the right
        // piece has length zero.
        let (left, right) = iter.split_at(index);
        debug_assert_eq!(right.len, 0);
        (
            (left, n_whole_chunks, last_dim.clone()),
            (right, 0, last_dim),
        )
    } else {
        // In this case, the right iterator contains the last chunk (and
        // possibly more chunks before it).
        let left_n_whole_chunks = index;
        let right_n_whole_chunks = n_whole_chunks - left_n_whole_chunks;
        let left_last_dim = iter.inner_dim.clone();
        let right_last_dim = last_dim;
        let (left, right) = iter.split_at(index);
        (
            (left, left_n_whole_chunks, left_last_dim),
            (right, right_n_whole_chunks, right_last_dim),
        )
    }
}
```

IMO, a cleaner way to implement this is to keep track not of `n_whole_chunks` but of the index corresponding to the partial chunk, and then implement `.split_at()` like this:

```rust
    pub fn split_at(self, index: usize) -> (Self, Self) {
        let (left, right) = self.iter.split_at(index);
        (
            AxisChunksIter {
                iter: left,
                partial_chunk_index: self.partial_chunk_index,
                partial_chunk_dim: self.partial_chunk_dim.clone(),
                life: self.life,
            },
            AxisChunksIter {
                iter: right,
                partial_chunk_index: self.partial_chunk_index,
                partial_chunk_dim: self.partial_chunk_dim,
                life: self.life,
            },
        )
    }
```

Once #669 is merged, I can provide a more complete suggestion.

Reply:
I've created #691 to add (hopefully correct) `.split_at()` implementations. Once that is merged, it should be straightforward to update this PR.

Reply:
@jturner314 great thanks! I'll update after that is merged.